File: java.py
import os
import subprocess

student_folder = "student_submissions"

def run_java(filepath):
    basename = os.path.splitext(os.path.basename(filepath))[0]
    try:
        compile_proc = subprocess.run(["javac", filepath], capture_output=True, text=True)
        if compile_proc.returncode != 0:
            return "compile_error", compile_proc.stderr
        run_proc = subprocess.run(["java", "-cp", os.path.dirname(filepath), basename], capture_output=True, text=True)
        if run_proc.returncode != 0:
            return "runtime_error", run_proc.stderr
        return "success", run_proc.stdout
    except Exception as e:
        return "exception", str(e)

grades = {}

for filename in os.listdir(student_folder):
    if filename.endswith(".java"):
        student_file = os.path.join(student_folder, filename)
        print(f"Grading {filename}...")
        status, output = run_java(student_file)
        if status == "success":
            grades[filename] = 100
        elif status == "compile_error":
            grades[filename] = 50
            print(f"Compilation Error in {filename}:\n{output}")
        elif status == "runtime_error":
            grades[filename] = 75
            print(f"Runtime Error in {filename}:\n{output}")
        else:
            grades[filename] = 0
            print(f"Unexpected Error in {filename}:\n{output}")

for fname, grade in grades.items():
    print(f"{fname}: {grade}")

File: java.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Grader {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the folder path where the submission files are: ");
        String submissionsFolder = sc.nextLine().trim();

        File folder = new File(submissionsFolder);
        if (!folder.exists() || !folder.isDirectory()) {
            System.out.println("Invalid folder path.");
            return;
        }

        Map<String, Integer> finalGrades = gradeSubmissions(folder);
        System.out.println("\nFinal Grades:");
        finalGrades.forEach((k, v) -> System.out.println(k + ": Grade = " + v));
    }

    public static Map<String, Integer> gradeSubmissions(File folder) {
        Map<String, Integer> grades = new LinkedHashMap<>();

        // Create a temp folder for grading one file at a time
        File tempDir = new File(folder, "_tempGrade");
        tempDir.mkdir();

        for (File file : Objects.requireNonNull(folder.listFiles())) {
            if (!file.getName().endsWith(".java")) continue;
            System.out.println("\nGrading " + file.getName() + "...");

            try {
                // Clean temp directory
                for (File f : Objects.requireNonNull(tempDir.listFiles())) f.delete();

                // Copy student's file into tempDir as OrderedSequentialSearchST.java
                Path dest = Paths.get(tempDir.getAbsolutePath(), "OrderedSequentialSearchST.java");
                Files.copy(file.toPath(), dest, StandardCopyOption.REPLACE_EXISTING);

                // Compile + run
                String status = runJavaFile(dest.toFile());
                int score = switch (status) {
                    case "success" -> 100;
                    case "runtime_error" -> 75;
                    case "compile_error" -> 50;
                    default -> 0;
                };
                grades.put(file.getName(), score);

            } catch (Exception e) {
                System.out.println("Error grading " + file.getName() + ": " + e.getMessage());
                grades.put(file.getName(), 0);
            }
        }

        // Clean up temp folder
        for (File f : Objects.requireNonNull(tempDir.listFiles())) f.delete();
        tempDir.delete();

        return grades;
    }

    private static String runJavaFile(File javaFile) {
        try {
            Process compile = new ProcessBuilder("javac", javaFile.getAbsolutePath())
                    .directory(javaFile.getParentFile())
                    .redirectErrorStream(true)
                    .start();
            String compileOut = new String(compile.getInputStream().readAllBytes());
            int compileExit = compile.waitFor();

            if (compileExit != 0) {
                System.out.println("Compilation Error:\n" + compileOut);
                return "compile_error";
            }

            Process run = new ProcessBuilder("java", "-cp", javaFile.getParent(), "OrderedSequentialSearchST")
                    .redirectErrorStream(true)
                    .start();
            String runOut = new String(run.getInputStream().readAllBytes());
            int runExit = run.waitFor();

            if (runExit != 0) {
                System.out.println("Runtime Error:\n" + runOut);
                return "runtime_error";
            }

            System.out.println("Output:\n" + runOut);
            return "success";
        } catch (Exception e) {
            System.out.println("Unexpected error: " + e.getMessage());
            return "exception";
        }
    }
}
